pkg/controller/storagecluster/placement_test.go

```diff
@@ -13,108 +13,105 @@ import (
-				Value:    "true",
-				Effect:   corev1.TaintEffectNoSchedule,
-			},
+var workerLabelSelector = metav1.LabelSelector{
+	MatchExpressions: []metav1.LabelSelectorRequirement{
+		metav1.LabelSelectorRequirement{
+			Key:      WorkerAffinityKey,
+			Operator: metav1.LabelSelectorOpExists,
 		},
 	},
 }
-var defaultLabelPlacement = map[rookv1.KeyType]rookv1.Placement{
-	"all": rookv1.Placement{
-		NodeAffinity: &corev1.NodeAffinity{
-			RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
-				NodeSelectorTerms: []corev1.NodeSelectorTerm{
-					corev1.NodeSelectorTerm{
-						MatchExpressions: []corev1.NodeSelectorRequirement{
-							corev1.NodeSelectorRequirement{
-								Key:      defaults.NodeAffinityKey,
-								Operator: corev1.NodeSelectorOpExists,
-							},
-						},
-					},
+var workerNodeSelector = corev1.NodeSelector{
+	NodeSelectorTerms: []corev1.NodeSelectorTerm{
+		corev1.NodeSelectorTerm{
+			MatchExpressions: []corev1.NodeSelectorRequirement{
+				corev1.NodeSelectorRequirement{
+					Key:      WorkerAffinityKey,
+					Operator: corev1.NodeSelectorOpExists,
 				},
 			},
 		},
-		Tolerations: []corev1.Toleration{
-			corev1.Toleration{
-				Key:      defaults.NodeTolerationKey,
-				Operator: corev1.TolerationOpEqual,
-				Value:    "true",
-				Effect:   corev1.TaintEffectNoSchedule,
+	},
+}
+
+var emptyLabelSelector = metav1.LabelSelector{
+	MatchExpressions: []metav1.LabelSelectorRequirement{},
+}
+var emptyNodeAffinity = corev1.NodeAffinity{
+	RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+		NodeSelectorTerms: []corev1.NodeSelectorTerm{
+			corev1.NodeSelectorTerm{
+				MatchExpressions: []corev1.NodeSelectorRequirement{},
 			},
 		},
 	},
 }
+var emptyPlacements = map[rookv1.KeyType]rookv1.Placement{
+	"all": rookv1.Placement{},
+}
 
 func TestGetPlacement(t *testing.T) {
-	// Case 1: Defaults are preserved i.e no placement and no label selector
-	sc := &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	assert.Equal(t, defaultLabelPlacement["all"], getPlacement(sc, "all"))
-
-	// Case 2: The configured Placements override the defaults
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.Placement = mockPlacements
-	assert.Equal(t, defaultLabelPlacement["all"], getPlacement(sc, "all"))
-
-	// Case 3: LabelSelector to modify the default placements correctly
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.LabelSelector = &metav1.LabelSelector{
-		MatchExpressions: []metav1.LabelSelectorRequirement{
-			metav1.LabelSelectorRequirement{
-				Key:      WorkerAffinityKey,
-				Operator: metav1.LabelSelectorOpExists,
+	cases := []struct {
+		placements        rookv1.PlacementSpec
+		labelSelector     *metav1.LabelSelector
+		expectedPlacement rookv1.Placement
+	}{
+		{
+			// Case 1: Defaults are preserved i.e no placement and no label selector
+			placements:    rookv1.PlacementSpec{},
```

Comment:
doesn't this mean "empty" placement and not "nil" (or "no" placement) ??

Reply:
I think it means placement has not been provided?

Reply:
Correct. Since the Placement field isn't a pointer, you can't set it to nil, you have to set it to an empty struct.

Reply:
Since the value being returned is from a list of non-pointers, it will never be nil. An empty struct is the only thing that can be returned if no placement was explicitly set.